Source/Core/Common/StringUtil.cpp

```diff
@@ -197,26 +197,41 @@ std::string StringFromFormatV(const char* format, va_list args)
   return temp;
 }
 
-// For Debugging. Read out an u8 array.
-std::string ArrayToString(const u8* data, u32 size, int line_len, bool spaces)
+// This assumes you already masked the upper four bits.
+constexpr char NibbleToHex(const u8 n)
 {
-  std::ostringstream oss;
-  oss << std::setfill('0') << std::hex;
+  if (n < 0xa)
+    return '0' + n;
+  else
+    return 'A' + n - 0xa;
+}
 
-  for (int line = 0; size; ++data, --size)
+std::string MemToHexString(const void* data, std::size_t size, const std::size_t per_line,
+                           const bool spaces)
+{
+  if (size == 0 || per_line == 0)
+    return {};
+  const u8* dataptr = reinterpret_cast<const u8*>(data);
+  const std::size_t nibble_count = size * 2;
+  const std::size_t newline_count = size / per_line - (size % per_line == 0);
+  const std::size_t space_count = (size - newline_count - 1) * spaces;
+  std::string hexstring(nibble_count + newline_count + space_count, ' ');
```

Comment:
Why preallocate the string here instead of just using `reserve()`, `push_back()`?

Reply:
It seems to be more optimal from the instruction count, but it is pretty hard to look at.